### PR TITLE
#113 - Implement experimental web worker support

### PR DESCRIPTION
**Pull request by treshugart**
Implements #40, #113.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -1 +0,0 @@
-require('skatejs-build');
```
Comment (treshugart) [inline gulpfile.js]: Totes unused.

File: test/unit.js [hunk 2]
```diff
@@ -465,3 +465,22 @@ describe('render', function () {
...
+describe('diff worker', function () {
+  it('should do work in a worker if opts.done is specified', done => {
+    const source = sd.vdom.element('div', null, 'text 1');
+    const destination = sd.vdom.element('div', null, 'text 2');
+    const realDom = sd.vdom.dom(source);
+
+    sd.diff({
+      destination,
+      source,
+      done (instructions) {
```
Comment (treshugart) [inline test/unit.js]: This is the tentatively proposed API: provide a `done` callback and if workers are available it uses them. Still need to patch `render` / `merge` to handle this.

File: src/util/real-node-map.js [hunk 3]
```diff
@@ -1,2 +1,11 @@
-import WeakMap from './weak-map';
-export default new WeakMap();
+// import WeakMap from './weak-map';
+// export default new WeakMap();
+const map = [];
+export default {
```
Comment (treshugart) [inline src/util/real-node-map.js]: Can't use weakmaps if we're using scalar values as keys.

File: src/index.js [hunk 4]
```diff
@@ -4,14 +4,12 @@ import patch from './patch';
...
-import version from './version';
 
 export default {
   diff,
   merge,
   patch,
   render,
   types,
-  vdom,
-  version
```
Comment (treshugart) [inline src/index.js]: Boyscout: removed this as it was way out of sync.

File: gulpfile.js (hunk 1 above)
Comment (matthewp) [inline gulpfile.js]: The build doesn't work now, is it because of this?

**Comment by treshugart** [comment]
Running sk-build should work.

**Comment by treshugart** [comment]
On mobile. Will try in an hour or so when on lappy.

**Comment by matthewp** [comment]
Hm, I don't have a `node_modules/.bin/sk-build`, is that something I need to install globally?

**Comment by treshugart** [comment]
Darn. Npm install should do that. Can't check atm but can in 45mins ish.

**Comment by matthewp** [comment]
Are you using both rollup and webpack? Or maybe transitioning from one to another?

**Comment by matthewp** [comment]
About the API, I think it's a good API for a lot of use cases, but in my case I'm already running inside of a web worker, so I don't need to launch another worker. I'll be diffing 2 plain objects in the worker, sending the patch instructions back to the window and applying the patches there.

To keep my file-size down I'll probably use rollup, does the use of the webpack plugin here break rollup usage (even if you don't wind up using that functionality)?  I could maybe just pull in dom-diff/src/diff-main.js directly since that's what I need on the worker since, then pull in dom-diff/src/patch.js on the window side.

**Comment by treshugart** [comment]
It should produce an artefact for the worker that you can load from dist
for diffing.

**Comment by treshugart** [comment]
Either way, yeah, you should be able to load the diff main in a worker,
that's what diff-worker does. You just need the deps that the worker-loader
gives you.

**Comment by treshugart** [comment]
Fixed the rollup build issue. Now generating a `dist/index.js` with webpack and will add a min version soon.

**Comment by matthewp** [comment]
So, I don't know if you answered this, but since you are using `jsnext` config, a rollup user might do:

```js
import { diff } from 'skatejs-dom-diff';
```

Will this work or blow up, aren't you using a webpack config in src/diff.js or am I just confused?

**Comment by treshugart** [comment]
You're right. I'll remove that. Travelling between Syd and Mel right now
which is why my responses have been patchy :) I'll have a look a bit later.

**Comment by matthewp** [comment]
ids don't seem to work, I guess because it assumes your nodes are created with `sd.vdom.element`, right? So that means it only works with patching 2 vdom nodes. You can't patch a real node against a vdom node or vice versa.

Also, if one Node is created in the worker and another in the window their ids aren't going to match up.  I'm pretty sure this is why React has ids that look like `0.1.14.2.3`, it's really a *path* to find the Node.  That's the only way to find the same Node from another tree.

I wrote a library for this purpose about a year ago: https://github.com/canjs/node-route/blob/master/src/dom-id.js  The inline docs do a decent job of explaining what is going on. Of particular interest are `nodeRoute.getID(node)` which gives you that id like "0.1.14.2.3". And other is `nodeRoute.getNode(id, root)` which will find the Node for a given id. Notice that it uses `firstChild` and `nextSibling` which your vdom doesn't have, so its not something you could drop in and use unfortunately. But could be a guide if you wanted to use it.

This could wind up being a bigger change that you were probably anticipating, I imagine, so I would understand if you don't want to pursue it further.

**Comment by treshugart** [comment]
> ids don't seem to work, I guess because it assumes your nodes are created with sd.vdom.element, right? So that means it only works with patching 2 vdom nodes. You can't patch a real node against a vdom node or vice versa.

This is intentional as there's no way to transfer a DOM node from the real DOM to a web worker. A way around this would be to extract information from each node and give it an id at that point in time. Unfortunately this would involve building a virtual tree from the entire real tree on the main thread. Basically the inverse of the `vdom/dom` function. This might be worth doing if performance isn't the main concern, which there's definitely use cases for, but if that's the case, then why use web workers?

What's the use case for needing to diff a real tree in a worker?

**Comment by matthewp** [comment]
You need to construct a virtual tree from the real tree once, when the page loads, if you ever intend to patch the real tree. I want to build the virtual tree on page load, send it to a worker, and from that point on only patch that virtual tree against other virtual trees, sending the patch instructions back to the window and applying the patches on the real tree.

**Comment by treshugart** [comment]
Massive update. I've added both `toDom()` and `toVdom()` to convert between real and virtual trees. You should be able to do something like:

```js
import { diff, toVdom } from 'skatejs-dom-diff';

const vTree = toVdom(document.getElementById('app'));
diff(vTree, myNewTree, {
  done (instructions) {
    patch(instructions);
  }
});
```

That said, the `render()` function now supports doing all that; so it will take the root node you provide it and convert it to a virtual tree before diffing and patching. It also supports the `done` function for diffing in a worker:

```js
import { render } from 'skatejs-dom-diff';

const renderer = render(root => (
  <div>I was rendered in {root.id}!</div>
));

renderer(document.getElementById('app'));
```

**Comment by matthewp** [comment]
Yeah no doubt, I think this is a real differentiator for the project, as I've explored all of them just about and most do not break out diffing and patching in this way.

I'm happy with idom but definitely am not married to it. I have plans for creating benchmarking for vdom through web workers approaches. In the meantime I still have a branch of my project using dom diff so I'll try out these changes and let you know.

**Pull request merged**